File: static/js/salary-formatter.js
/**
 * Salary Formatting Utilities for Frontend
 * Provides consistent salary formatting with currency across all templates
 */

class SalaryFormatter {
    /**
     * Format salary range with proper currency display
     * @param {number|null} salaryLow - Minimum salary amount
     * @param {number|null} salaryHigh - Maximum salary amount  
     * @param {string} currency - Currency code (CAD, USD, etc.)
     * @returns {string} Formatted salary string with currency
     */
    static formatRange(salaryLow = null, salaryHigh = null, currency = 'CAD') {
        if (!salaryLow && !salaryHigh) {
            return "Salary not specified";
        }
        
        if (salaryLow && salaryHigh) {
            if (salaryLow === salaryHigh) {
                return `$${salaryLow.toLocaleString()} ${currency}`;
            } else {
                return `$${salaryLow.toLocaleString()} - $${salaryHigh.toLocaleString()} ${currency}`;
            }
        } else if (salaryLow) {
            return `$${salaryLow.toLocaleString()}+ ${currency}`;
        } else if (salaryHigh) {
            return `Up to $${salaryHigh.toLocaleString()} ${currency}`;
        }
        
        return "Salary not specified";
    }

    /**
     * Format single salary amount with currency and period
     * @param {number} amount - Salary amount
     * @param {string} currency - Currency code (CAD, USD, etc.)
     * @param {string} period - Salary period (annually, monthly, hourly)
     * @returns {string} Formatted salary string
     */
    static formatSingle(amount, currency = 'CAD', period = 'annually') {
        if (!amount) {
            return "Amount not specified";
        }
        
        let formatted = `$${amount.toLocaleString()} ${currency}`;
        
        if (period && period !== 'annually') {
            formatted += ` ${period}`;
        }
        
        return formatted;
    }

    /**
     * Determine currency based on location
     * @param {string} location - Location string
     * @param {string} country - Country name
     * @returns {string} Currency code (CAD, USD, etc.)
     */
    static getCurrencyFromLocation(location = null, country = null) {
        if (country) {
            const countryLower = country.toLowerCase();
            if (countryLower.includes('canada')) {
                return 'CAD';
            } else if (countryLower.includes('united states') || countryLower.includes('usa')) {
                return 'USD';
            }
        }
        
        if (location) {
            const locationLower = location.toLowerCase();
            // Canadian indicators
            if (locationLower.includes('canada') || 
                locationLower.includes('alberta') || 
                locationLower.includes('ontario') ||
                locationLower.includes('quebec') ||
                locationLower.includes('bc') ||
                locationLower.includes('british columbia') ||
                locationLower.includes('saskatchewan') ||
                locationLower.includes('manitoba') ||
                locationLower.includes('nova scotia') ||
                locationLower.includes('new brunswick') ||
                locationLower.includes('newfoundland') ||
                locationLower.includes('pei') ||
                locationLower.includes('yukon') ||
                locationLower.includes('northwest territories') ||
                locationLower.includes('nunavut')) {
                return 'CAD';
            }
            // US indicators  
            else if (locationLower.includes('usa') ||
                     locationLower.includes('united states') ||
                     locationLower.includes('california') ||
                     locationLower.includes('texas') ||
                     locationLower.includes('new york') ||
                     locationLower.includes('florida')) {
                return 'USD';
            }
        }
        
        // Default to CAD for Canadian job board
        return 'CAD';
    }

    /**
     * Parse salary text and extract components
     * @param {string} salaryText - Raw salary text
     * @returns {object} Object with salary_low, salary_high, currency, period
     */
    static parseSalaryText(salaryText) {
        const result = {
            salary_low: null,
            salary_high: null,
            currency: 'CAD',
            period: 'annually'
        };
        
        if (!salaryText) {
            return result;
        }
        
        const text = salaryText.toString();
        
        // Extract currency
        if (text.toUpperCase().includes('USD')) {
            result.currency = 'USD';
        } else if (text.toUpperCase().includes('CAD')) {
            result.currency = 'CAD';
        }
        
        // Extract period
        const textLower = text.toLowerCase();
        if (textLower.includes('hour') || textLower.includes('hourly') || textLower.includes('/hr') || textLower.includes('per hour')) {
            result.period = 'hourly';
        } else if (textLower.includes('month') || textLower.includes('monthly') || textLower.includes('/mo') || textLower.includes('per month')) {
            result.period = 'monthly';
        }
        
        // Extract numbers
        const numbers = text.replace(/,/g, '').match(/\d+(?:\.\d+)?/g);
        if (numbers) {
            const nums = numbers.map(n => parseInt(parseFloat(n)));
            if (nums.length === 1) {
                result.salary_low = result.salary_high = nums[0];
            } else if (nums.length >= 2) {
                result.salary_low = Math.min(...nums);
                result.salary_high = Math.max(...nums);
            }
        }
        
        return result;
    }

    /**
     * Update job display elements with proper salary formatting
     * @param {HTMLElement} container - Container element with job data
     * @param {object} jobData - Job data object
     */
    static updateJobDisplays(container, jobData) {
        const salaryElements = container.querySelectorAll('[data-salary]');
        
        salaryElements.forEach(element => {
            const currency = jobData.salary_currency || this.getCurrencyFromLocation(jobData.location, jobData.country);
            const formattedSalary = this.formatRange(jobData.salary_low, jobData.salary_high, currency);
            element.textContent = formattedSalary;
        });
    }
}

// Make available globally
window.SalaryFormatter = SalaryFormatter;

// Convenience functions for backwards compatibility
window.formatSalaryRange = SalaryFormatter.formatRange;
window.formatSingleSalary = SalaryFormatter.formatSingle;
window.getCurrencyFromLocation = SalaryFormatter.getCurrencyFromLocation;

File: modules/salary_formatter.py
"""
Salary Formatting Utilities
Provides consistent salary formatting with currency across the application
"""

from typing import Optional


def format_salary_range(
    salary_low: Optional[int] = None, salary_high: Optional[int] = None, currency: str = "CAD"
) -> str:
    """
    Format salary range with proper currency display

    Args:
        salary_low: Minimum salary amount
        salary_high: Maximum salary amount
        currency: Currency code (CAD, USD, etc.)

    Returns:
        Formatted salary string with currency
    """
    if not salary_low and not salary_high:
        return "Salary not specified"

    if salary_low and salary_high:
        if salary_low == salary_high:
            return f"${salary_low:,} {currency}"
        else:
            return f"${salary_low:,} - ${salary_high:,} {currency}"
    elif salary_low:
        return f"${salary_low:,}+ {currency}"
    elif salary_high:
        return f"Up to ${salary_high:,} {currency}"

    return "Salary not specified"


def format_single_salary(amount: Optional[int], currency: str = "CAD", period: str = "annually") -> str:
    """
    Format single salary amount with currency and period

    Args:
        amount: Salary amount
        currency: Currency code (CAD, USD, etc.)
        period: Salary period (annually, monthly, hourly)

    Returns:
        Formatted salary string
    """
    if not amount:
        return "Amount not specified"

    formatted_amount = f"${amount:,} {currency}"

    if period and period != "annually":
        formatted_amount += f" {period}"

    return formatted_amount


def get_currency_from_location(location: Optional[str] = None, country: Optional[str] = None) -> str:
    """
    Determine currency based on location

    Args:
        location: Location string
        country: Country name

    Returns:
        Currency code (CAD, USD, etc.)
    """
    if country:
        if "canada" in country.lower():
            return "CAD"
        elif "united states" in country.lower() or "usa" in country.lower():
            return "USD"

    if location:
        location_lower = location.lower()
        # Canadian indicators
        if any(
            indicator in location_lower
            for indicator in [
                "canada",
                "alberta",
                "ontario",
                "quebec",
                "bc",
                "british columbia",
                "saskatchewan",
                "manitoba",
                "nova scotia",
                "new brunswick",
                "newfoundland",
                "pei",
                "yukon",
                "northwest territories",
                "nunavut",
            ]
        ):
            return "CAD"
        # US indicators
        elif any(
            indicator in location_lower
            for indicator in ["usa", "united states", "california", "texas", "new york", "florida"]
        ):
            return "USD"

    # Default to CAD for Canadian job board
    return "CAD"


def extract_salary_from_text(salary_text: str) -> dict:
    """
    Extract salary information from text with currency detection

    Args:
        salary_text: Raw salary text

    Returns:
        Dictionary with salary_low, salary_high, currency, period
    """
    import re

    result = {"salary_low": None, "salary_high": None, "currency": "CAD", "period": "annually"}

    if not salary_text:
        return result

    # Extract currency
    if "USD" in salary_text.upper():
        result["currency"] = "USD"
    elif "CAD" in salary_text.upper():
        result["currency"] = "CAD"

    # Extract period
    text_lower = salary_text.lower()
    if any(word in text_lower for word in ["hour", "hourly", "/hr", "per hour"]):
        result["period"] = "hourly"
    elif any(word in text_lower for word in ["month", "monthly", "/mo", "per month"]):
        result["period"] = "monthly"

    # Extract numbers
    numbers = re.findall(r"\d+(?:\.\d+)?", salary_text.replace(",", ""))
    if numbers:
        nums = [int(float(n)) for n in numbers]
        if len(nums) == 1:
            result["salary_low"] = result["salary_high"] = nums[0]
        elif len(nums) >= 2:
            result["salary_low"] = min(nums)
            result["salary_high"] = max(nums)

    return result
